src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Types.cs

```diff
@@ -19,31 +19,21 @@ private void VisitTypeSymbolWithAnnotations(TypeSymbolWithAnnotations type, Abst
             bool includeNullability = format.MiscellaneousOptions.IncludesOption(SymbolDisplayMiscellaneousOptions.IncludeNullableReferenceTypeModifier);
             bool includeNonNullability = format.CompilerInternalOptions.IncludesOption(SymbolDisplayCompilerInternalOptions.IncludeNonNullableTypeModifier);
 
-            // Don't un-necessarily pull on IsNullable to avoid cycles in parts of binding that use ToDisplayString
-            bool? isNullable = (includeNullability || includeNonNullability) ? type.IsNullable : null;
-
-            if (typeSymbol.TypeKind == TypeKind.Array && isNullable.HasValue)
+            if (typeSymbol.TypeKind == TypeKind.Array)
             {
-                visitor.VisitArrayType((IArrayTypeSymbol)typeSymbol, isNullable: isNullable);
+                visitor.VisitArrayType((IArrayTypeSymbol)typeSymbol, isNullable: type);
             }
             else
             {
                 typeSymbol.Accept(visitor);
 
-                switch (isNullable)
+                if (includeNonNullability && !typeSymbol.IsNullableType() && type.IsAnnotated)
```

Comment:
`if (includeNullability && …)`?

Reply:
Thanks. Hold off on detailed review. I'm going to rebase this shortly. Also still working on some failing tests.
I'll stop by tomorrow to sync on this PR (solved problems, remaining problems).